plutus-core/executables/plutus/AnyProgram/Compile.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeOperators #-}
+module AnyProgram.Compile
+    ( compileProgram
+    ) where
+
+import Types
+import AnyProgram.With
+
+import PlutusCore.Default
+import PlutusCore qualified as PLC
+import PlutusCore.DeBruijn qualified as PLC
+import PlutusCore.Compiler qualified as PLC
+import PlutusIR qualified as PIR
+import PlutusIR.Error qualified as PIR
+import PlutusIR.Compiler qualified as PIR
+import UntypedPlutusCore qualified as UPLC
+import UntypedPlutusCore.Check.Uniques qualified as UPLC
+
+import PlutusPrelude hiding ((%~))
+import Data.Singletons.Decide
+import Control.Monad.Reader
+import Control.Lens hiding ((%~))
+import Control.Monad.Except
+import Control.Monad.Error.Lens
+
+compileProgram :: (e ~ (PIR.Provenance (FromAnn (US_ann s1))), -- Here we use the erroring of the original term's annotation
+                  MonadError (PIR.Error DefaultUni DefaultFun e) m)
+               => SLang s1 -> SLang s2 -> FromLang s1 -> m (FromLang s2)
+compileProgram = curry $ \case
+    -- exclude all pir-debruijn input&output
+    (SPir SNamedDeBruijn _, _) -> const $ throwing PIR._Error $ PIR.OptionsError "pir input cannot be debruijn"
+    (SPir SDeBruijn _, _ ) -> const $ throwing PIR._Error $ PIR.OptionsError "pir input cannot be nameddebruijn"
+    (_, SPir SDeBruijn _) -> const $ throwing PIR._Error $ PIR.OptionsError "pir out cannot be debruijn"
+    (_, SPir SNamedDeBruijn _) -> const $ throwing PIR._Error $ PIR.OptionsError "pir out cannot be nameddebruijn"
+
+    -- pir to plc
+    (sng1@(SPir n1@SName a1), SPlc n2 a2@SUnit) ->
+        -- prove that a1 has Ord,Pretty
+        withOrdPrettyA a1 $
+                (PLC.runQuoteT . flip runReaderT (PIR.toDefaultCompilationCtx undefined) . PIR.compileProgram)
+            -- get rid of annotations for now
+            >=> pure . void
+            >=> plcToOutName n1 n2
+    (SPir SName _, SPlc _ _) -> const $ throwing PIR._Error $ PIR.OptionsError "only support unit-ann output for now"
```

Comment:
It should be straightforward to allow output ann type to be the same as the input ann type, isn't it?